Validator failed migration to 0.4.0

[ This issue was auto-migrated from DA's internal repo (DACH-NY/canton-network-node#19761). Original author: @isegall-da ]

[logs.tar.gz](https://github.com/user-attachments/files/20506689/logs.tar.gz)

From validator logs:

Successful ACS import:
```

2025-05-29T13:45:00.226Z [⋮] INFO - o.l.s.m.DomainDataRestorer:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - Importing the ACS                                                                                 

2025-05-29T13:45:00.428Z [⋮] DEBUG - o.l.s.a.a.c.ApiClientRequestLogger:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - Request (tid:14a95199f1ce5494cc706dd67c6a1dbe) com.digitalasset.canton.admin.participant.v30.ParticipantRepairService/ImportAcsOld to participant:5002: sending request                                                                                                                                                                                

2025-05-29T13:45:06.370Z [⋮] DEBUG - o.l.s.a.a.c.ApiClientRequestLogger:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - Request (tid:14a95199f1ce5494cc706dd67c6a1dbe) com.digitalasset.canton.admin.participant.v30.ParticipantRepairService/ImportAcsOld to participant:5002: received a message                                                                                                                                                                             

2025-05-29T13:45:06.453Z [⋮] DEBUG - o.l.s.a.a.c.ApiClientRequestLogger:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - Request (tid:14a95199f1ce5494cc706dd67c6a1dbe) com.digitalasset.canton.admin.participant.v30.ParticipantRepairService/ImportAcsOld to participant:5002: succeeded(OK)                                                                                                                                                                                  

2025-05-29T13:45:06.479Z [⋮] INFO - o.l.s.m.DomainDataRestorer:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - Imported the ACS                                                                                  
```

The participant agrees:
```
2025-05-29T13:45:05.600Z [⋮] INFO - c.d.c.p.i.IndexerState:participant=participant (---) - Committing repair changes succeeded, resuming normal indexing...
2025-05-29T13:45:06.340Z [⋮] DEBUG - c.d.c.n.g.ApiRequestLogger:participant=participant (14a95199f1ce5494cc706dd67c6a1dbe---) - Request com.digitalasset.canton.admin.participant.v30.ParticipantRepairService/ImportAcsOld by /10.228.1.46:33536: succeeded(OK)
```


Losing the participant:

```
2025-05-29T13:45:14.481Z [⋮] INFO - o.l.s.e.ParticipantAdminConnection:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - reconnect to the domain Synchronizer 'global' for new sequencer configuration to take effect

2025-05-29T13:46:17.303Z [⋮] INFO - o.l.s.e.ParticipantAdminConnection:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-connect_domain-6fd3f96211422079-c4a278ce273afb81) - The operation 'participant is connected to Synchronizer 'global'' failed with a retryable error (full stack trace omitted): UNAVAILABLE: Network closed for unknown reason

2025-05-29T13:46:22.589Z [⋮] ERROR - o.l.s.v.ValidatorApp:validator=validator_backend (ecabcbdd0fb279a450159b32a06917b0-app_init--6fd3f96211422079) - validator_backend app initialization: Initialization failed java.lang.RuntimeException: validator_backend app initialization: Initialize app failed
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
```

But nothing in the participant logs on why, it seems to be doing its thing until 13:46:16, then disappearing and restarting at 13:46:43:
```
2025-05-29T13:46:16.521Z [⋮] DEBUG - c.d.c.t.p.TopologyTransactionAuthorizationValidator:participant=participant/synchronizerId=global-domain::122071b3b1bd/role=incoming (e3ad67dfc1645bab7889dee9654bf538---) - Required authorizations: RequiredNamespaces(Set(122071b3b1bd..., 1220875f7f4f...))                                                                                                                                                                                                                       
2025-05-29T13:46:16.522Z [⋮] DEBUG - c.d.c.t.p.TopologyTransactionAuthorizationValidator:participant=participant/synchronizerId=global-domain::122071b3b1bd/role=incoming (e3ad67dfc1645bab7889dee9654bf538---) - Authorizations for namespaces: 1220875f7f4f... fully authorized by keys Set(1220875f7f4f...), 122071b3b1bd... fully authorized by keys Set(12208412efbf..., 1220963511a3..., 122069eedf7c...)                                                                                                            
2025-05-29T13:46:16.543Z [⋮] DEBUG - c.d.c.t.p.TopologyTransactionAuthorizationValidator:participant=participant/synchronizerId=global-domain::122071b3b1bd/role=incoming (e3ad67dfc1645bab7889dee9654bf538---) - All keys used for authorization: Set(1220875f7f4f..., 12208412efbf..., 1220963511a3..., 122069eedf7c...)                                                                                                                                                                                                 2025-05-29T13:46:43.000Z [⋮] INFO - json_log (---) - Starting '/app/bin/canton' with arguments: daemon --no-tty --log-encoder=json --log-level-stdout=DEBUG --log-level-canton=DEBUG --log-file-appender=off --config /app/app.conf --config /app/monitoring.conf --config /app/parameters.conf                                                                                                                                                                                                                            2025-05-29T13:46:45.738Z [⋮] INFO - c.d.c.CantonEnterpriseApp$ (---) - Starting Canton version 3.3.0-SNAPSHOT                                                                                                                                                
```

Trying another ACS import after restarting, fails because participant is lost again:
```
2025-05-29T13:49:49.138Z [⋮] INFO - o.l.s.m.DomainDataRestorer:validator=validator_backend (9e0e9478876342f4f04ee36c220e8c27-app_init--957163cfe5f0bb50) - Importing the ACS

┌2025-05-29T13:50:49.681Z [⋮] INFO - o.l.s.e.ParticipantAdminConnection:validator=validator_backend (9e0e9478876342f4f04ee36c220e8c27-import_acs-957163cfe5f0bb50-961ea10ed1558356) - The operation 'Imports the acs in the participantl' failed with a retryable error (full stack trace omitted): DEADLINE_EXCEEDED: CallOptions deadline exceeded after 59.962597549s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[remote_addr=participant/34.118.233.10:5002]]]

2025-05-29T13:57:54.928Z [⋮] ERROR - o.l.s.v.ValidatorApp:validator=validator_backend (9e0e9478876342f4f04ee36c220e8c27-app_init--957163cfe5f0bb50) - validator_backend app initialization: Initialization failed java.lang.RuntimeException: validator_backend app initialization: Initialize app failed
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
```

Again nothing in the participant logs on why:

```
2025-05-29T13:57:51.130Z [⋮] INFO - c.d.c.p.a.r.RepairService:participant=participant (584c46d76bdb09e5c3e62444b214c8d5---) - Task repair.add still not completed after 4 minutes. Continue waiting...                                                       
2025-05-29T13:57:51.455Z [⋮] INFO - c.d.c.p.a.r.RepairService:participant=participant (baaecf630f7314d6ad6b55f1a9b2653b---) - Task repair.add still not completed after 8 minutes. Continue waiting...                                                       
2025-05-29T13:57:51.752Z [⋮] DEBUG - c.d.c.p.a.r.RepairService:participant=participant (584c46d76bdb09e5c3e62444b214c8d5---) - repair.add has not completed after 4 minutes                                                                                  
2025-05-29T13:57:51.752Z [⋮] DEBUG - c.d.c.p.a.r.RepairService:participant=participant (baaecf630f7314d6ad6b55f1a9b2653b---) - repair.add has not completed after 8 minutes                                                                                  
2025-05-29T13:57:53.000Z [⋮] INFO - json_log (---) - Starting '/app/bin/canton' with arguments: daemon --no-tty --log-encoder=json --log-level-stdout=DEBUG --log-level-canton=DEBUG --log-file-appender=off --config /app/app.conf --config /app/monitoring.conf --config /app/parameters.conf                                                                                                                                                                                                                            2025-05-29T13:57:58.189Z [⋮] INFO - c.d.c.CantonEnterpriseApp$ (---) - Starting Canton version 3.3.0-SNAPSHOT                                                                                                                                                
2025-05-29T13:57:58.221Z [⋮] DEBUG - c.d.c.CantonEnterpriseApp$ (---) - Registered shutdown-hook.                                                                                                                                                            
2025-05-29T13:58:03.463Z [⋮] INFO - c.d.c.CantonEnterpriseApp$ (---) - Starting up with resolved config                                                                                                                                                      
```

Trying another ACS import after restarting, fails on import error:
```
2025-05-29T14:01:32.794Z [⋮] INFO - o.l.s.m.DomainDataRestorer:validator=validator_backend (7112bba9a7dbb66d4f675294a8a0d1a1-app_init--5255ee357d2b7154) - Importing the ACS

025-05-29T14:01:35.110Z [⋮] INFO - o.l.s.a.a.c.ApiClientRequestLogger:validator=validator_backend (7112bba9a7dbb66d4f675294a8a0d1a1-app_init--5255ee357d2b7154) - Request (tid:1eb728fc56ae54bc8ef03dec56528d52) com.digitalasset.canton.admin.participant.v30.ParticipantRepairService/ImportAcsOld to participant:5002: failed with INVALID_ARGUMENT/IMPORT_ACS_ERROR(8,1eb728fc): Cannot add previously archived contract ContractId(00f7153bdba834f04ec9c4e17f323f2a2b615c8769bd22e7bf175ff6e3d67f2645ca1012207874bc623cf8015f156e78cbf290d568d312044bf679682921d285926672996e) as archived contracts cannot become active.
```

--- a/apps/common/src/main/scala/org/lfdecentralizedtrust/splice/environment/SpliceLedgerConnection.scala
+++ b/apps/common/src/main/scala/org/lfdecentralizedtrust/splice/environment/SpliceLedgerConnection.scala
@@ -1240,8 +1240,6 @@ object BaseLedgerConnection {
 
   val SV_NAME_USER_METADATA_KEY: String = "sv.app.network.canton.global/sv_name"
 
-  val INITIAL_ACS_IMPORT_METADATA_KEY: String = "network.canton.global/initial_acs_import"
-
   val SV1_INITIAL_PACKAGE_UPLOAD_METADATA_KEY: String =
     "network.canton.global/sv1_initial_package_upload"
 

--- a/apps/common/src/main/scala/org/lfdecentralizedtrust/splice/migration/DomainDataRestorer.scala
+++ b/apps/common/src/main/scala/org/lfdecentralizedtrust/splice/migration/DomainDataRestorer.scala
@@ -3,11 +3,7 @@
 
 package org.lfdecentralizedtrust.splice.migration
 
-import org.lfdecentralizedtrust.splice.environment.{
-  BaseLedgerConnection,
-  ParticipantAdminConnection,
-  RetryFor,
-}
+import org.lfdecentralizedtrust.splice.environment.{ParticipantAdminConnection, RetryFor}
 import org.lfdecentralizedtrust.splice.util.UploadablePackage
 import com.digitalasset.canton.config.{SynchronizerTimeTrackerConfig, NonNegativeFiniteDuration}
 import com.digitalasset.canton.logging.{NamedLoggerFactory, NamedLogging}
@@ -30,8 +26,6 @@ class DomainDataRestorer(
   /** We assume the domain was not register prior to trying to restore the data.
     */
   def connectDomainAndRestoreData(
-      ledgerConnection: BaseLedgerConnection,
-      userId: String,
       synchronizerAlias: SynchronizerAlias,
       synchronizerId: SynchronizerId,
       sequencerConnections: SequencerConnections,
@@ -40,52 +34,53 @@ class DomainDataRestorer(
   )(implicit
       tc: TraceContext
   ): Future[Unit] = {
-    logger.info("Registering and connecting to new domain")
+    def restoreData() = {
+      val domainConnectionConfig = SynchronizerConnectionConfig(
+        synchronizerAlias,
+        synchronizerId = Some(synchronizerId),
+        sequencerConnections = sequencerConnections,
+        manualConnect = true,
+        initializeFromTrustedSynchronizer = true,
+        timeTracker = SynchronizerTimeTrackerConfig(
+          timeTrackerMinObservationDuration
+        ),
+      )
+      // We rely on the calls here being idempotent
+      for {
+        // Disconnect
+        _ <- participantAdminConnection.disconnectFromAllDomains()
+        _ <- importDars(dars)
+        _ = logger.info("Imported all the dars.")
+        _ <-
+          participantAdminConnection
+            .ensureDomainRegistered(
+              domainConnectionConfig,
+              RetryFor.ClientCalls,
+            )
+        _ = logger.info("Importing the ACS")
+        _ <- importAcs(acsSnapshot)
+        _ = logger.info("Imported the ACS")
+        _ <- participantAdminConnection.modifySynchronizerConnectionConfigAndReconnect(
+          synchronizerAlias,
+          config => Some(config.copy(manualConnect = false)),
+        )
+      } yield ()
+    }
 
     // We use user metadata as a dumb storage to track whether we already imported the ACS.
-    ledgerConnection
-      .lookupUserMetadata(
-        userId,
-        BaseLedgerConnection.INITIAL_ACS_IMPORT_METADATA_KEY,
+    participantAdminConnection
+      .lookupSynchronizerConnectionConfig(
+        synchronizerAlias
       )
       .flatMap {
         case None =>
-          val domainConnectionConfig = SynchronizerConnectionConfig(
-            synchronizerAlias,
-            synchronizerId = Some(synchronizerId),
-            sequencerConnections = sequencerConnections,
-            manualConnect = true,
-            initializeFromTrustedSynchronizer = true,
-            timeTracker = SynchronizerTimeTrackerConfig(
-              timeTrackerMinObservationDuration
-            ),
+          logger.info("Synchronizer not yet registered, registering and restoring data")
+          restoreData()
+        case Some(conf) if conf.manualConnect == true =>
+          logger.info(
+            "Synchronizer registered but manualConnect=true, assuming we crashed during a prior attempt and trying again"
           )
-          // We rely on the calls here being idempotent
-          for {
-            // Disconnect
-            _ <- participantAdminConnection.disconnectFromAllDomains()
-            _ <- importDars(dars)
-            _ = logger.info("Imported all the dars.")
-            _ <-
-              participantAdminConnection
-                .ensureDomainRegistered(
-                  domainConnectionConfig,
-                  RetryFor.ClientCalls,
-                )
-            _ = logger.info("Importing the ACS")
-            _ <- importAcs(acsSnapshot)
-            _ = logger.info("Imported the ACS")
-            _ <- participantAdminConnection.modifySynchronizerConnectionConfigAndReconnect(
-              synchronizerAlias,
-              config => Some(config.copy(manualConnect = false)),
-            )
-            _ <- ledgerConnection.ensureUserMetadataAnnotation(
-              userId,
-              BaseLedgerConnection.INITIAL_ACS_IMPORT_METADATA_KEY,
-              "true",
-              RetryFor.ClientCalls,
-            )
-          } yield ()
+          restoreData()
         case Some(_) =>
           logger.info("Domain is already registered and ACS is imported")
           participantAdminConnection.connectDomain(synchronizerAlias)

--- a/apps/sv/src/main/scala/org/lfdecentralizedtrust/splice/sv/onboarding/domainmigration/DomainMigrationInitializer.scala
+++ b/apps/sv/src/main/scala/org/lfdecentralizedtrust/splice/sv/onboarding/domainmigration/DomainMigrationInitializer.scala
@@ -265,8 +265,6 @@ class DomainMigrationInitializer(
         ),
       )
       _ <- domainDataRestorer.connectDomainAndRestoreData(
-        readOnlyConnection,
-        config.ledgerApiUser,
         synchronizerAlias,
         domainMigrationDump.nodeIdentities.synchronizerId,
         SequencerConnections.single(localSynchronizerNode.sequencerConnection),

--- a/apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/ValidatorApp.scala
+++ b/apps/validator/src/main/scala/org/lfdecentralizedtrust/splice/validator/ValidatorApp.scala
@@ -226,8 +226,6 @@ class ValidatorApp(
                       loggerFactory,
                     )
                     decentralizedSynchronizerInitializer.connectDomainAndRestoreData(
-                      connection,
-                      config.ledgerApiUser,
                       config.domains.global.alias,
                       migrationDump.domainId,
                       sequencerConnections,

--- a/docs/src/release_notes.rst
+++ b/docs/src/release_notes.rst
@@ -16,6 +16,12 @@ Upcoming
   - Fix a bug where sends through transfer preapprovals failed with a ``CONTRACT_NOT_FOUND`` ERROR
     if the receiver's provider party was featured.
 
+- Synchronizer Migrations
+
+  - Fix a rare bug where a crash of the validator or SV while trying
+    to restore the data after a migration could result in an
+    inconsistent state being restore.
+
 0.4.2
 -----
 

